MULE-19423: Adding description of the mule.http.requestStreaming.enable property (#1712)
mulesoft/docs-connectors: http/1.6/modules/ROOT/pages/http-request-ref.adoc

By default, the body of a request is taken from the `#[payload]` of the incoming Mule message and the response is sent onwards as the `#[payload]` of the output Mule message. You can change this default behavior through the *General > Request > Body* and *General > Output > Target Variable* attributes. Use this attribute to specify a location other than payload for the output data, such as a variable.

== Configure Request Streaming Mode

> When HTTP Connector manages request bodies, the connector considers the type of data to send. You can configure the HTTP *Request* operation to send HTTP requests in chunks. By default, if the payload type is stream, the operation uses the streaming mechanism to send HTTP requests. +

Reviewer: What are "request bodies": payload?
What are the "type of data to send"; are these different types of responses? What types?

> When streaming, the HTTP request does not contain the `Content-Length` header. The request contains the `Transfer-Encoding` header and sends the body in chunks until the stream is fully consumed.

Reviewer: Is this a level of detail that the audience can understand here at the beginning of the topic?
How does this relate to the previous text?

> To change the default behavior, configure *Request streaming mode* to any of these options:

Reviewer: This is misleading, because the second option is NOT a change of the default behavior; it IS the default behavior.

So I wonder if this topic should actually be Change the Default Request Streaming Mode, and begin by saying that that by default, the connector uses Auto mode, but that you can change it (if you want to achieve <this?>
And then list and define modes.

* *ALWAYS* +
Always enable streaming regardless of the payload type.
* *AUTO* (default) +

> The behavior depends on the payload type. If the payload is stream, streaming is enabled, otherwise it is disabled.

Reviewer 2: Should "stream" have back ticks?

> The behavior depends on the payload type. If the payload is stream, streaming is enabled, otherwise it is disabled.

Reviewer 2: ```suggestion
The behavior depends on the payload type. If the payload is a stream, streaming is enabled. Otherwise, it is disabled.
```

Reviewer: agreed

* *NEVER* +
Never stream, even if the payload is a stream.

> The following example shows how to configure  *Request streaming mode* in Studio:

Reviewer: ```suggestion
The following example shows how to configure *ALWAYS* request streaming mode in Studio:
```

. In the *Mule Palette* view, select *HTTP > Request*.
. Drag *Request* to the Studio canvas.
. Set *Path* to `/path`.
. In the HTTP *Request* operation configuration screen, click the *Request* tab.
. Set *Request streaming mode* to `ALWAYS`.

image::http-request-6.png[HTTP Request configuration window with Response streaming mode field set to ALWAYS]

> In the Configuration XML editor, the `requestStreamingMode` configuration looks like this:

Reviewer: Is Configuration capped because it is part of the editor's name? If it is just a type, lowercase; if a label, use bold formatting.

[source,xml,linenums]
----
<flow name="httprequestFlow"  >
		<http:request method="GET" requestStreamingMode="ALWAYS" config-ref="HTTP_Request_configuration" path="/path"/>
	</flow>
----

> === Configure Request Streaming System Property

Reviewer 2: ```suggestion
=== Configure the Request Streaming System Property
```

> The previous *Request streaming mode* configuration causes the Mule application to send HTTP requests in chunks, but the configuration doesn't imply that the Mule app streams the payload internally. The HTTP *Request* operation needs all the data in memory before sending HTTP requests. Therefore, if the Mule app needs to send a huge request body, the Mule app can generate a high memory consumption, and eventually cause an out-of-memory error. To prevent such an issue, Mule provides the `mule.http.requestStreaming.enable` system property to enable internal request streaming.

Reviewer 2: Inconsistent term - app or application

Reviewer 2: ```suggestion
The previous *Request streaming mode* configuration causes the Mule application to send HTTP requests in chunks, but the configuration doesn't imply that the Mule app streams the payload internally. The HTTP *Request* operation needs all the data in memory before sending HTTP requests. Therefore, if the Mule app needs to send a huge request body, the Mule app can generate a high memory consumption, and eventually cause an out-of-memory error. To prevent this issue, Mule provides the `mule.http.requestStreaming.enable` system property to enable internal request streaming.
```

Reviewer 2: I suggest changing "huge" to "very large". To me "huge" sounds too dramatic.

> The previous *Request streaming mode* configuration causes the Mule application to send HTTP requests in chunks, but the configuration doesn't imply that the Mule app streams the payload internally. The HTTP *Request* operation needs all the data in memory before sending HTTP requests. Therefore, if the Mule app needs to send a huge request body, the Mule app can generate a high memory consumption, and eventually cause an out-of-memory error. To prevent such an issue, Mule provides the `mule.http.requestStreaming.enable` system property to enable internal request streaming.

Reviewer: Instead of "the previous," can we call it "the example configuration of *Request streaming mode* to `Always`..."?

Reviewer: Yeah...I just don't get the technology here, I'm sorry.
I'm looking for a definition of the difference between configuring the streaming mode vs. the streaming system property.
And it still seems odd to me that streams are sent in chunks, because what is the alternative (a chunk seems the alternative to me).
Stick with Diane's copyedits here.

> If you configure the system property, the HTTP *Request* operation won't wait until having the whole request body send the request, and then the Mule app won't need that memory.

Reviewer 2: ```suggestion
If you configure this system property, the HTTP *Request* operation starts sending the request body before all of the data is in memory. Therefore, the Mule app won't need as much memory as it would need if you don't configure this property.
```

Reviewer 2: If you want to keep more of the original wording:

Reviewer: If you configure this system property, the HTTP *Request* operation starts sending the request body before all of the data is in memory, resulting in the Mule app consuming less memory.

> To configure the system property, set it when starting Mule by adding `-M-Dmule.http.requestStreaming.enable=true`. The system property defaults to `false`.

Reviewer: ```suggestion
Configure this system property when you start Mule by changing the default value `-M-Dmule.http.requestStreaming.enable=false` to `-M-Dmule.http.requestStreaming.enable=true`.

> You can also configure the internal buffer `mule.http.requestStreaming.bufferSize` system property for the streaming mechanism. To configure the system property, set it when starting Mule by adding `-M-Dmule.http.requestStreaming.bufferSize=value`. The system property is measured in bytes and defaults to `8192`.

Reviewer 2: ```suggestion
You can also configure the internal buffer `mule.http.requestStreaming.bufferSize` system property for the streaming mechanism. To configure this property, set it when starting Mule by adding `-M-Dmule.http.requestStreaming.bufferSize=value`. The `-M-Dmule.http.requestStreaming.bufferSize=value` system property is measured in bytes and defaults to `8192`.
```

Reviewer: Is this an additional way to save memory or an alternative?
Also, I think you can shorten to 
"You can also configure the internal buffer `mule.http.requestStreaming.bufferSize` system property for the streaming mechanism by adding `-M-Dmule.http.requestStreaming.bufferSize=value`, where `value` is a number of bytes and defaults to `8192`.
